legacy-sdk/whirlpool/src/utils/math/bit-math.ts

```diff
@@ -64,7 +66,7 @@ export class BitMath {
       );
     }
     const result = MathUtil.fromX64_BN(p);
-    const shouldRound = roundUp && result.and(U64_MAX).gt(ZERO);
+    const shouldRound = roundUp && p.and(U64_MAX).gt(ZERO);
```

Comment:
fix existing bug (result is already shifted, so result.and(U64_MAX) is almost always greater than zero.
(math in the contract-side is safe 🙂)

Reply:
ah good catch